code/network/lib/Loot/Network/ZMQ/Common.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE RecordWildCards #-}
+
+-- | Common types and functions.
+
+module Loot.Network.ZMQ.Common
+    ( ZmqTcp
+    , ZTGlobalEnv(..)
+    , ztContext
+    , withZTGlobalEnv
+    , endpointTcp
+    , ZTNodeId(..)
+    , ztNodeIdRouter
+    , ztNodeIdPub
+    , ztNodeConnectionId
+    ) where
+
+
+import Codec.Serialise (Serialise)
+import Control.Lens (makeLenses)
+import qualified Data.ByteString.Char8 as BS8
+
+import qualified Data.Restricted as Z
+import qualified System.ZMQ4 as Z
+
+
+
+-- | Networking tag type for ZMQ over TCP.
+data ZmqTcp
+
+data ZTGlobalEnv = ZTGlobalEnv
+    { _ztContext :: Z.Context
+    }
+
+makeLenses ''ZTGlobalEnv
+
+withZTGlobalEnv :: (MonadMask m, MonadIO m) => (ZTGlobalEnv -> m a) -> m a
+withZTGlobalEnv action =
+    bracket (liftIO Z.context) (liftIO . Z.term) $ action . ZTGlobalEnv
+
+-- | Generic tcp address creation helper.
+endpointTcp :: String -> Integer -> String
+endpointTcp h p = "tcp://" <> h <> ":" <> show p
+
+data ZTNodeId = ZTNodeId
+    { ztIdHost       :: String
+    , ztIdRouterPort :: Integer
+    , ztIdPubPort    :: Integer
+    } deriving (Eq, Ord, Show, Generic)
+
+instance Serialise ZTNodeId
+
+-- | Address of the server's ROUTER/frontend socket.
+ztNodeIdRouter :: ZTNodeId -> String
+ztNodeIdRouter ZTNodeId{..} = endpointTcp ztIdHost ztIdRouterPort
```

Comment:
I'm against usage of `RecordWildCards`. The `NamedFieldPuns` at least does not pop a bunch of names from nowhere.